### DISCUSSION TITLE
How to test a non-public class used as a type parameter (this is not about InternalsVisibleTo)?

### DISCUSSION THREAD

**Discussion by IanKemp** [xunit/xunit#2659] [Question] [answered]
We have a company-internal NuGet package that we use to simplify our unit testing which exposes a `ClassContext<TClass>` class that uses AutoFixture and Moq to automagically mock all dependencies of `TClass`, unless those are specifically overridden. It's used as such:

```csharp
// in main project (with InternalsVisibleTo)
internal class MyClassUnderTest
{
    ...
}

// in test project
public class MyXUnitTestClass : ClassContext<MyClassUnderTest>
{
    ...
}
```

The problem: `MyClassUnderTest` is `internal` [but XUnit requires test classes to be public](https://xunit.net/xunit.analyzers/rules/xUnit1000).
The result: [Compiler error CS0060](https://learn.microsoft.com/dotnet/csharp/misc/cs0060) as `MyXUnitTestClass` and `MyClassUnderTest`
The fix: make `MyClassUnderTest` `public`.
The problem: `MyClassUnderTest` is intentionally `internal` because it shouldn't be consumable by anything outside its assembly.

This is extremely sub-optimal because XUnit is essentially forcing us to modify our code to allow us to test it. There are ways to work around this by changing our code (e.g. to have an instance of `ClassContext<T>` instead of inheriting from it) but I'd prefer to not have to do that, because again that would be working around XUnit, not with it.

So I'd like to ask:

* is there a way to accomplish the scenario demonstrated in the code above without making **any** changes to the class under test, and without making major changes to the test class?
* if not, **why** does the XUnit limitation on test classes being `public` exist, and is there a way to work around it? If not, would a feature request to remove this limitation be considered? If not, why not?

**Comment by bradwilson** [comment, marked as answer]
I'm afraid there is no workaround for this, and none planned. The only way you will be able to make this compile is to use some mechanism other than inheritance, since it's the inheritance that is causing the compiler error.

**Reply by IanKemp** [reply]
What is the reason XUnit does not support non-`public` test classes, and why don't you want to change this?

**Reply by bradwilson** [reply]
For performance reasons: reflecting over all private types is extremely expensive.
